include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,148 @@
+        };
+
+    /*
+     * Type erasure class for allocators. This allows to use any kind of allocator
+     * (standard, polymorphic) whithout having to expose it as a template parameter.
+     *
+     * @tparam T value_type of the allocator
+     * @tparam DA default allocator, instantiated when calling the default constructor
+     */
+    template <class T, class DA = std::allocator<T>>
+    class any_allocator
+    {
+    public:
+
+        using value_type = T;
+
+        any_allocator()
+            : p_allocator(std::make_unique<impl<DA>>(DA()))
+        {
+        }
+
+        any_allocator(const any_allocator& rhs)
+            : p_allocator(rhs.p_allocator->clone())
+        {
+        }
+
+        any_allocator& operator=(const any_allocator& rhs)
+        {
+            p_allocator = rhs.p_allocator->clone();
+            return *this;
+        }
+
+        any_allocator(any_allocator&&) = default;
+        any_allocator& operator=(any_allocator&&) = default;
+
+        template <class A>
+        any_allocator(A&& alloc)
+        requires (not std::same_as<std::remove_cvref_t<A>, any_allocator>
+                and sparrow::allocator<A>)
+            : p_allocator(std::make_unique<impl<std::decay_t<A>>>(std::forward<A>(alloc)))
+        {
+        }
+
+        [[nodiscard]] T* allocate(std::size_t n)
+        {
+            return p_allocator->allocate(n);
+        }
+
+        void deallocate(T* p, std::size_t n)
+        {
+            p_allocator->deallocate(p, n);
+        }
+
+        bool equal(const any_allocator& rhs) const
+        {
+            return p_allocator->equal(*rhs.p_allocator);
+        }
+
+    private:
+
+        struct interface
+        {
+            [[nodiscard]] virtual T* allocate(std::size_t) = 0;
+            virtual void deallocate(T*, std::size_t) = 0;
+            virtual std::unique_ptr<interface> clone() const = 0;
+            virtual bool equal(const interface&) const = 0;
+            virtual ~interface() = default;
+        };
+
+        template <allocator A>
+        struct impl : interface
+        {
+            A m_alloc;
+
+            explicit impl(A alloc) : m_alloc(std::move(alloc)) {}
+
+            [[nodiscard]] T* allocate(std::size_t n) override
+            {
+                return m_alloc.allocate(n);
+            }
+
+            void deallocate(T* p, std::size_t n) override
+            {
+                m_alloc.deallocate(p, n);
+            }
+
+            std::unique_ptr<interface> clone() const override
+            {
+                return std::make_unique<impl<A>>(m_alloc);
+            }
+
+            bool equal(const interface& rhs) const override
+            {
+                if (auto* p = dynamic_cast<const impl<A>*>(&rhs))
+                {
+                    return p->m_alloc == m_alloc;
+                }
+                return false;
+            }
+        };
+
+        std::unique_ptr<interface> p_allocator;
```

Comment:
While we agreed to start with minimal effort for this implementation, I believe we might want to immediately support either small buffer optimization (there are various ways to do this but maybe using `std::variant`, `std::any` or `std::function` - the last 2 ones already do sbo) OR just support at least that there is no allocation (no `make_unique`) when the allocator type is actually the default (`DA`) by then just storing it as a value (maybe using a `variant<D1, unique_ptr>`?).

My main worry is the cost in "normal" case where `DA` is `std::allocator`.

Reply:
Indeed, let's fix this issue before this gets in.